core/src/client/mod.rs

```diff
@@ -29,11 +28,12 @@ pub struct Client<N: NetworkSpec, C: Consensus<N::BlockResponse>> {
 impl<N: NetworkSpec, C: Consensus<N::BlockResponse>> Client<N, C> {
     pub fn new(
         execution_rpc: &str,
+        verifiable_api: Option<&str>,
```

Comment:
I don't love that we have both the execution rpc and verifiable api address as inputs. I wonder if there is some way to pass one or the other?

Maybe an enum or alternatively we can construct an the api/rpc client before and pass it in?

Reply:
I think that's only possible if passing verifiable_api means you don't need to pass execution_rpc. But that is not the case in the current implementation since the former doesn't completely replace the latter.

Reply:
What methods do we need to fall back on the execution rpc to verify? We should just have versions of those added to the verifiable api (even if they don't really change much from the rpc) so that we can have a single endpoint that helios consumes.

Reply:
All the methods where response needs to be verified are there in the VerifiableApi (see PR description for list). I've not implemented methods where we just need to proxy the same request/response without any change.

These would include methods like `eth_getBlockReceipts`, `eth_newFilter`, `eth_newBlockFilter` and many more that are used by Helios but also methods like `eth_getBlockTransactionCountByHash`, `eth_getBlockTransactionCountByNumber` and many more which are not used by Helios and rather constructed from Helios state.

We can think about implementing the ones that are used to replace the RPC but I don't see any point in the ones that are not used at all.

Reply:
Yeah I agree no need to implement unused methods, but we should ensure that helios can consume either data source as its sole data source rather than needing to rely on both. I think it will only require adding one or two more endpoints.

Reply:
Will need to add the following additional RPC methods:

- `eth_getBlockByHash`
- `eth_getBlockByNumber`
- `eth_chainId`
- `eth_sendRawTransaction`
- `eth_getBlockReceipts`
- `eth_newFilter`
- `eth_newBlockFilter`
- `eth_newPendingTransactionFilter`
- `eth_uninstallFilter`

Shouldn't be too much effort to implement since we just wanna proxy the same request/response.

I'll get started on this?

Reply:
Sounds good

Reply:
Implemented endpoints for these methods in 236a9052564f2832bc0d6780071538511cb4657a.

Reply:
There are two other places in the codebase that directly use the HttpRpc instead of ExecutionClient.

- https://github.com/a16z/helios/blob/b4cabb52732c797f71613d98e1f73ea63cf1aed3/opstack/src/server/mod.rs#L92
- https://github.com/a16z/helios/blob/b4cabb52732c797f71613d98e1f73ea63cf1aed3/core/src/execution/state.rs#L324

I am not sure how we can replace this with ExecutionClient since that also requires `state`.

Reply:
Need to encapsulate this better if we want the behavior that user only needs to specify either verifiable_api URL or execution_rpc URL but not both.

Reply:
For the server this shouldn't matter as this is backend infra for our opstack integration, and never runs on the client.

As for the state, we should definitely handle that using either the rpc or the api.

Reply:
This is now solved in 9864ba5da34caa81c2036cfba771f8c0ea37c7b5.